# Machine Learning Engineer Nanodegree
## Supervised Learning
## Project: Building a Student Intervention System

Welcome to the second project of the Machine Learning Engineer Nanodegree! In this notebook, some template code has already been provided for you, and it will be your job to implement the additional functionality necessary to successfully complete this project. Sections that begin with **'Implementation'** in the header indicate that the following block of code will require additional functionality which you must provide. Instructions will be provided for each section and the specifics of the implementation are marked in the code block with a `'TODO'` statement. Please be sure to read the instructions carefully!

In addition to implementing code, there will be questions that you must answer which relate to the project and your implementation. Each section where you will answer a question is preceded by a **'Question X'** header. Carefully read each question and provide thorough answers in the following text boxes that begin with **'Answer:'**. Your project submission will be evaluated based on your answers to each of the questions and the implementation you provide.  

>**Note:** Code and Markdown cells can be executed using the **Shift + Enter** keyboard shortcut. In addition, Markdown cells can be edited by typically double-clicking the cell to enter edit mode.

### Question 1 - Classification vs. Regression
*Your goal for this project is to identify students who might need early intervention before they fail to graduate. Which type of supervised learning problem is this, classification or regression? Why?*

**Answer: **

classification problem

For regrassion problem, it mostly deal with continouse inputs, with outputs as continous numbers. 
In this case here, lots of inputs features are discrete, such as boolean (e.g. "paid": yes, no), categorical data(e.g."guardian": "father","mother","other"). Meanwhile, the expected result is boolen: "Yes, need early intervention" and "No, don't need intervention", it is also discrete. In the case, classfication is preferable.

## Exploring the Data
Run the code cell below to load necessary Python libraries and load the student data. Note that the last column from this dataset, `'passed'`, will be our target label (whether the student graduated or didn't graduate). All other columns are features about each student.

In [65]:
# Import libraries
import numpy as np
import pandas as pd
from time import time
from sklearn.metrics import f1_score

# Read student data
student_data = pd.read_csv("student-data.csv")
print "Student data read successfully!"

Student data read successfully!


### Implementation: Data Exploration
Let's begin by investigating the dataset to determine how many students we have information on, and learn about the graduation rate among these students. In the code cell below, you will need to compute the following:
- The total number of students, `n_students`.
- The total number of features for each student, `n_features`.
- The number of those students who passed, `n_passed`.
- The number of those students who failed, `n_failed`.
- The graduation rate of the class, `grad_rate`, in percent (%).


In [66]:
# TODO: Calculate number of students
n_students = len(student_data.axes[0])

# TODO: Calculate number of features
n_features = len(student_data.columns[:-1])
# n_features = len(student_data.axes[1])

# TODO: Calculate passing students
n_passed = len(student_data[student_data.passed == "yes"])

# TODO: Calculate failing students
n_failed = len(student_data[student_data.passed == "no"])

# TODO: Calculate graduation rate
grad_rate = n_passed / float(n_passed + n_failed)

# Print the results
print "Total number of students: {}".format(n_students)
print "Number of features: {}".format(n_features)
print "Number of students who passed: {}".format(n_passed)
print "Number of students who failed: {}".format(n_failed)
print "Graduation rate of the class: {:.2f}%".format(grad_rate * 100)

Total number of students: 395
Number of features: 30
Number of students who passed: 265
Number of students who failed: 130
Graduation rate of the class: 67.09%


## Preparing the Data
In this section, we will prepare the data for modeling, training and testing.

### Identify feature and target columns
It is often the case that the data you obtain contains non-numeric features. This can be a problem, as most machine learning algorithms expect numeric data to perform computations with.

Run the code cell below to separate the student data into feature and target columns to see if any features are non-numeric.

In [67]:
# Extract feature columns
feature_cols = list(student_data.columns[:-1])

# Extract target column 'passed'
target_col = student_data.columns[-1] 

# Show the list of columns
print "Feature columns:\n{}".format(feature_cols)
print "\nTarget column: {}".format(target_col)

# Separate the data into feature data and target data (X_all and y_all, respectively)
X_all = student_data[feature_cols]
y_all = student_data[target_col]

# Show the feature information by printing the first five rows
print "\nFeature values:"
print X_all.head()

Feature columns:
['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']

Target column: passed

Feature values:
  school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

    ...    higher internet  romantic  famrel  freetime goout Dalc Walc health  \
0   ...       yes       no        no       4         3     4    1    1      3   
1   ...       

### Preprocess Feature Columns

As you can see, there are several non-numeric columns that need to be converted! Many of them are simply `yes`/`no`, e.g. `internet`. These can be reasonably converted into `1`/`0` (binary) values.

Other columns, like `Mjob` and `Fjob`, have more than two values, and are known as _categorical variables_. The recommended way to handle such a column is to create as many columns as possible values (e.g. `Fjob_teacher`, `Fjob_other`, `Fjob_services`, etc.), and assign a `1` to one of them and `0` to all others.

These generated columns are sometimes called _dummy variables_, and we will use the [`pandas.get_dummies()`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html?highlight=get_dummies#pandas.get_dummies) function to perform this transformation. Run the code cell below to perform the preprocessing routine discussed in this section.

In [68]:
def preprocess_features(X):
    ''' Preprocesses the student data and converts non-numeric binary variables into
        binary (0/1) variables. Converts categorical variables into dummy variables. '''
    
    # Initialize new output DataFrame
    output = pd.DataFrame(index = X.index)

    # Investigate each feature column for the data
    for col, col_data in X.iteritems():
        
        # If data type is non-numeric, replace all yes/no values with 1/0
        if col_data.dtype == object:
            col_data = col_data.replace(['yes', 'no'], [1, 0])

        # If data type is categorical, convert to dummy variables
        if col_data.dtype == object:
            # Example: 'school' => 'school_GP' and 'school_MS'
            col_data = pd.get_dummies(col_data, prefix = col)  
        
        # Collect the revised columns
        output = output.join(col_data)
    
    return output

X_all = preprocess_features(X_all)
print "Processed feature columns ({} total features):\n{}".format(len(X_all.columns), list(X_all.columns))

Processed feature columns (48 total features):
['school_GP', 'school_MS', 'sex_F', 'sex_M', 'age', 'address_R', 'address_U', 'famsize_GT3', 'famsize_LE3', 'Pstatus_A', 'Pstatus_T', 'Medu', 'Fedu', 'Mjob_at_home', 'Mjob_health', 'Mjob_other', 'Mjob_services', 'Mjob_teacher', 'Fjob_at_home', 'Fjob_health', 'Fjob_other', 'Fjob_services', 'Fjob_teacher', 'reason_course', 'reason_home', 'reason_other', 'reason_reputation', 'guardian_father', 'guardian_mother', 'guardian_other', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']


### Implementation: Training and Testing Data Split
So far, we have converted all _categorical_ features into numeric values. For the next step, we split the data (both features and corresponding labels) into training and test sets. In the following code cell below, you will need to implement the following:
- Randomly shuffle and split the data (`X_all`, `y_all`) into training and testing subsets.
  - Use 300 training points (approximately 75%) and 95 testing points (approximately 25%).
  - Set a `random_state` for the function(s) you use, if provided.
  - Store the results in `X_train`, `X_test`, `y_train`, and `y_test`.

In [69]:
# TODO: Import any additional functionality you may need here
from sklearn.cross_validation import train_test_split

# TODO: Set the number of training points
num_train = 300

# Set the number of testing points
num_test = X_all.shape[0] - num_train

test_size = num_test / float(X_all.shape[0])

# TODO: Shuffle and split the dataset into the number of training and testing points above
X_train, X_test, y_train, y_test = train_test_split(X_all, y_all, test_size = test_size, random_state=42)




# Show the results of the split
print "Training set has {} samples.".format(X_train.shape[0])
print "Testing set has {} samples.".format(X_test.shape[0])

Training set has 300 samples.
Testing set has 95 samples.


## Training and Evaluating Models
In this section, you will choose 3 supervised learning models that are appropriate for this problem and available in `scikit-learn`. You will first discuss the reasoning behind choosing these three models by considering what you know about the data and each model's strengths and weaknesses. You will then fit the model to varying sizes of training data (100 data points, 200 data points, and 300 data points) and measure the F<sub>1</sub> score. You will need to produce three tables (one for each model) that shows the training set size, training time, prediction time, F<sub>1</sub> score on the training set, and F<sub>1</sub> score on the testing set.

**The following supervised learning models are currently available in** [`scikit-learn`](http://scikit-learn.org/stable/supervised_learning.html) **that you may choose from:**
- Gaussian Naive Bayes (GaussianNB)
- Decision Trees
- Ensemble Methods (Bagging, AdaBoost, Random Forest, Gradient Boosting)
- K-Nearest Neighbors (KNeighbors)
- Stochastic Gradient Descent (SGDC)
- Support Vector Machines (SVM)
- Logistic Regression

### Question 2 - Model Application
*List three supervised learning models that are appropriate for this problem. For each model chosen*
- Describe one real-world application in industry where the model can be applied. *(You may need to do a small bit of research for this — give references!)* 
- What are the strengths of the model; when does it perform well? 
- What are the weaknesses of the model; when does it perform poorly?
- What makes this model a good candidate for the problem, given what you know about the data?

**Answer: **


*GaussianNB

Application: to mark an email as spam or not spam; 

Advantages: GaussianNB classifier is simple and easy to apply. If the naive assumption holds, NB classifier could converge faster than discriminative models; even if the assumption doesn’t hold, the NB classifier has surprisingly good performance in real practice. 
Perform well: when attributes are independent from each other, and there are only two hypothesis

Weakness: assumption of independent attributes is too constraining (though performs surprising well sometimes). NB classifier does not allow rich hypothesis. 
Perform poorly: when there are attributes highly dependent to each other, or there are large number of hypothesis. 

Why for this case: in this case, there are only two hypothesis, student need intervention, and student does not need intervention. Most features are independent from each other. Also, the dataset is small (396 rows X 48 rows), in which situation high bias/low variance classifiers like GaussianNB are preferable. 


*K-nearest Neighbor

Industry application: Recommender system, if you always purchase one item, you might  be interested in similar items

Advantages: easy to use, easy to train, high practical accuracy.
Perform well: when the data has small number of dimensions

Weakness: take lots of space to run, good for small number of dimensions, not for a high number of dimensions
Perform poorly: when the dataset has large number of dimensions 
 
Why for this case: the dataset in this case is comparatively small dimension, also, K-NN would provide with high accuracy.


*Decision Trees

Application: in banks, a decision tree to decide whether or not offer a credit card invitation

Advantages: easy to understand and interpret, deal well with little hard data piece, allow the addition of new scenarios, could decide the best and wrost scenario, could combined with other model. Overall, Decision Trees classifier could work directly with a table of features without very much preprocessing work.
Perform well: it performs well when features are all categorical and not with many levels,   and the number of features are not overwhelming

Weakness: easy to overfit, need prune the tree. Sensitive to small pertubation in the data, little changes in data could lead to a drastically different tree.
Perform wrong: large number of features, value under features are continuous data and hard to categorize them.

Why for this case: in this case, most features are categorical or could be combined as categorical.  



### Setup
Run the code cell below to initialize three helper functions which you can use for training and testing the three supervised learning models you've chosen above. The functions are as follows:
- `train_classifier` - takes as input a classifier and training data and fits the classifier to the data.
- `predict_labels` - takes as input a fit classifier, features, and a target labeling and makes predictions using the F<sub>1</sub> score.
- `train_predict` - takes as input a classifier, and the training and testing data, and performs `train_clasifier` and `predict_labels`.
 - This function will report the F<sub>1</sub> score for both the training and testing data separately.

In [70]:
def train_classifier(clf, X_train, y_train):
    ''' Fits a classifier to the training data. '''
    
    # Start the clock, train the classifier, then stop the clock
    start = time()
    clf.fit(X_train, y_train)
    end = time()
    
    # Print the results
    print "Trained model in {:.4f} seconds".format(end - start)

    
def predict_labels(clf, features, target):
    ''' Makes predictions using a fit classifier based on F1 score. '''
    
    # Start the clock, make predictions, then stop the clock
    start = time()
    y_pred = clf.predict(features)
    end = time()
    
    # Print and return results
    print "Made predictions in {:.4f} seconds.".format(end - start)
    return f1_score(target.values, y_pred, pos_label='yes')


def train_predict(clf, X_train, y_train, X_test, y_test):
    ''' Train and predict using a classifer based on F1 score. '''
    
    # Indicate the classifier and the training set size
    print "Training a {} using a training set size of {}. . .".format(clf.__class__.__name__, len(X_train))
    
    # Train the classifier
    train_classifier(clf, X_train, y_train)
    
    # Print the results of prediction for both training and testing
    print "F1 score for training set: {:.4f}.".format(predict_labels(clf, X_train, y_train))
    print "F1 score for test set: {:.4f}.".format(predict_labels(clf, X_test, y_test))

### Implementation: Model Performance Metrics
With the predefined functions above, you will now import the three supervised learning models of your choice and run the `train_predict` function for each one. Remember that you will need to train and predict on each classifier for three different training set sizes: 100, 200, and 300. Hence, you should expect to have 9 different outputs below — 3 for each model using the varying training set sizes. In the following code cell, you will need to implement the following:
- Import the three supervised learning models you've discussed in the previous section.
- Initialize the three models and store them in `clf_A`, `clf_B`, and `clf_C`.
 - Use a `random_state` for each model you use, if provided.
 - **Note:** Use the default settings for each model — you will tune one specific model in a later section.
- Create the different training set sizes to be used to train each model.
 - *Do not reshuffle and resplit the data! The new training points should be drawn from `X_train` and `y_train`.*
- Fit each model with each training set size and make predictions on the test set (9 in total).  
**Note:** Three tables are provided after the following code cell which can be used to store your results.

In [71]:
# TODO: Import the three supervised learning models from sklearn
from sklearn.cross_validation import train_test_split
# from sklearn import model_A
from sklearn.naive_bayes import GaussianNB
# from sklearn import model_B
from sklearn.tree import DecisionTreeClassifier
# from skearln import model_C
from sklearn.neighbors import KNeighborsClassifier as KNN

# TODO: Initialize the three models
clf_A = GaussianNB()
clf_B = DecisionTreeClassifier(random_state = 0)
clf_C = KNN()

# from sklearn.svm import SVC
# clf_D = SVC()
    
def train_model_size(model, num_train):
    X_train_num = X_train[- num_train:]
    y_train_num = y_train[- num_train:]
    train_predict(model, X_train_num, y_train_num, X_test, y_test)
    print "#####################"

train_model_size(clf_A, 100)
train_model_size(clf_A, 200)
train_model_size(clf_A, 300)

train_model_size(clf_B, 100)
train_model_size(clf_B, 200)
train_model_size(clf_B, 300)

train_model_size(clf_C, 100)
train_model_size(clf_C, 200)
train_model_size(clf_C, 300)

# train_model_size(clf_D, 100)
# train_model_size(clf_D, 200)
# train_model_size(clf_D, 300)

Training a GaussianNB using a training set size of 100. . .
Trained model in 0.0007 seconds
Made predictions in 0.0011 seconds.
F1 score for training set: 0.8308.
Made predictions in 0.0006 seconds.
F1 score for test set: 0.6720.
#####################
Training a GaussianNB using a training set size of 200. . .
Trained model in 0.0069 seconds
Made predictions in 0.0010 seconds.
F1 score for training set: 0.8102.
Made predictions in 0.0004 seconds.
F1 score for test set: 0.7258.
#####################
Training a GaussianNB using a training set size of 300. . .
Trained model in 0.0014 seconds
Made predictions in 0.0006 seconds.
F1 score for training set: 0.8038.
Made predictions in 0.0004 seconds.
F1 score for test set: 0.7634.
#####################
Training a DecisionTreeClassifier using a training set size of 100. . .
Trained model in 0.0009 seconds
Made predictions in 0.0003 seconds.
F1 score for training set: 1.0000.
Made predictions in 0.0002 seconds.
F1 score for test set: 0.6721.
##

### Tabular Results
Edit the cell below to see how a table can be designed in [Markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#tables). You can record your results from above in the tables provided.

** Classifer 1 - GaussianNaiveBayes (GaussianNB)**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |            0.0008          |     0.0004                 |        0.8308          |          0.6720      |
| 200               |        0.001          |             0.0006           |         0.8102         |         0.7258       |
| 300               |          0.001               |           0.0004           |       0.8038           |    0.7634     |

** Classifer 2 - DecisionTreeClassifier**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |     0.0011                 |      0.0002              |     1             |          0.6721       |
| 200               |     0.0013      |             0.0003          |          1        |         0.7519     |
| 300               |          0.0029              |          0.0002              |     1             |     0.6102     |

** Classifer 3 - K-Nearest NeighborsClassifier (KNN)**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |         0.0005              |             0.001          |       0.8322         |         0.7737     |
| 200               |          0.0006              |            0.0015    |        0.8882         |0.7626 |
| 300               |     0.0008                   |       0.0029                |    0.8809             |        0.7801         |

## Choosing the Best Model
In this final section, you will choose from the three supervised learning models the *best* model to use on the student data. You will then perform a grid search optimization for the model over the entire training set (`X_train` and `y_train`) by tuning at least one parameter to improve upon the untuned model's F<sub>1</sub> score. 

### Question 3 - Choosing the Best Model
*Based on the experiments you performed earlier, in one to two paragraphs, explain to the board of supervisors what single model you chose as the best model. Which model is generally the most appropriate based on the available data, limited resources, cost, and performance?*

**Answer: **



Regarding F1 scores for testing set, KNN gets averagely highest F1 score. In KNN, the highest score is when testing size is 300. There are no significant variance over different training set size (0.76-0.78). GaussianNB gets the averagely 0.72 F1 score, the higher the training set size, the higher F1 score. There are comparatively higher F1 score variances according to training set. DecisionTree gets the lowest average F1 score for testing set, 0.68, and has highest F1 score 0.75 when training set size is 200. There are comparatively highest F1 score variances among different training set sizes. 

Regarding F1 scores for training set, DecisionTree gets the highest average score, 1. KNN gets the second highest average, 0.86, and GaussianNB gets the lowest average score 0.82.

Considering the computing cost, KNN get shortest training time for each training size, GaussianNB is a little bit higher except when training size is 100. DecisionTree gets much higher training time. For prediction, GaussianNB and DecisionTree performs a little bit better job. KNN gets much higher training time with small training size. The average predition time for KNN is 5-10 times that for GaussianNB and DecisionTree.

In conclusion, KNN gets the best performance, with low variances according to different training size, and the performance will be improved as training size increases. Thought the prediciton time for KNN is higher than the other two, but the differences are acceptable. Overall, K-nearest Neighbour is the best model for this practice. 

### Question 4 - Model in Layman's Terms
*In one to two paragraphs, explain to the board of directors in layman's terms how the final model chosen is supposed to work. Be sure that you are describing the major qualities of the model, such as how the model is trained and how the model makes a prediction. Avoid using advanced mathematical or technical jargon, such as describing equations or discussing the algorithm implementation.*

**Answer: **


- The KNN model is "trained" by simply storing information about previous students who passed or failed. These "features" include info such as age, gender, absences, etc.
- To predict an outcome for a new student, we compare his/her information with that from those previous students, with respect to the value of each feature.
- The comparison returns a measure of how "similar" the new student and the previous student are, for example, new student with age = 10, gender = female is more close to previous student with age = 10, gender = female, compare to previous student with age = 5, gender = male.
- We then take the "k" most similar previous students and look for whether a majority of them passed or failed ("k" is a predefined number that you could always change, e.g. K = 1, 2, 3...)
- If a majority of the nearest neighbor students passed, we predict that the new student would pass, since he/she has similar features with neighbor students, he/she would has the similar result. 
- The whole idea of KNN is close to the saying "The birds of a feather gather together".

### Implementation: Model Tuning
Fine tune the chosen model. Use grid search (`GridSearchCV`) with at least one important parameter tuned with at least 3 different values. You will need to use the entire training set for this. In the code cell below, you will need to implement the following:
- Import [`sklearn.grid_search.GridSearchCV`](http://scikit-learn.org/0.17/modules/generated/sklearn.grid_search.GridSearchCV.html) and [`sklearn.metrics.make_scorer`](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html).
- Create a dictionary of parameters you wish to tune for the chosen model.
 - Example: `parameters = {'parameter' : [list of values]}`.
- Initialize the classifier you've chosen and store it in `clf`.
- Create the F<sub>1</sub> scoring function using `make_scorer` and store it in `f1_scorer`.
 - Set the `pos_label` parameter to the correct value!
- Perform grid search on the classifier `clf` using `f1_scorer` as the scoring method, and store it in `grid_obj`.
- Fit the grid search object to the training data (`X_train`, `y_train`), and store it in `grid_obj`.

In [72]:
# TODO: Import 'GridSearchCV' and 'make_scorer'
from sklearn.grid_search import GridSearchCV
from sklearn.metrics import f1_score, make_scorer

# TODO: Create the parameters list you wish to tune
parameters = {"n_neighbors": [1, 3, 5, 8, 10, 15], "weights": ["uniform", "distance"], 'algorithm':['ball_tree','kd_tree','brute'],
              'leaf_size':[5, 8, 10, 15, 20, 25, 30, 35, 40, 45, 50]}

# TODO: Initialize the classifier
clf = KNN()

# # ‘linear’, ‘poly’, ‘rbf’, ‘sigmoid’, ‘precomputed’
# parameters = {"C": [1, 3, 5, 8, 10], "kernel": ["poly", "rbf", "sigmoid", "linear"]}
# clf = SVC()

# TODO: Make an f1 scoring function using 'make_scorer' 
f1_scorer = make_scorer(f1_score, pos_label = "yes")

# TODO: Perform grid search on the classifier using the f1_scorer as the scoring method
grid_obj = GridSearchCV(clf, parameters, f1_scorer)

# TODO: Fit the grid search object to the training data and find the optimal parameters
grid_obj.fit(X_train, y_train)

# Get the estimator
clf = grid_obj.best_estimator_

# Report the final F1 score for training and testing after parameter tuning
print "Tuned model has a training F1 score of {:.4f}.".format(predict_labels(clf, X_train, y_train))
print "Tuned model has a testing F1 score of {:.4f}.".format(predict_labels(clf, X_test, y_test))


Made predictions in 0.0029 seconds.
Tuned model has a training F1 score of 1.0000.
Made predictions in 0.0013 seconds.
Tuned model has a testing F1 score of 0.7815.


### Question 5 - Final F<sub>1</sub> Score
*What is the final model's F<sub>1</sub> score for training and testing? How does that score compare to the untuned model?*

**Answer: **

Final model:
F1 score (training): 1
F1 score (testing): 0.7815

The untuned KNN model, F1 score for training set is 0.7815, for testing set is 0.7801. Compare final model with untuned model, the train F1 score got improved, but no huge differences for testing F1 score.


> **Note**: Once you have completed all of the code implementations and successfully answered each question above, you may finalize your work by exporting the iPython Notebook as an HTML document. You can do this by using the menu above and navigating to  
**File -> Download as -> HTML (.html)**. Include the finished document along with this notebook as your submission.